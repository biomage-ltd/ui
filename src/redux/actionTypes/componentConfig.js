/**
 * Sets embedding categorical config.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;

export { LOAD_CONFIG, UPDATE_CONFIG, SAVE_CONFIG };
