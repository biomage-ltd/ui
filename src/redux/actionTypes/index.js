const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
