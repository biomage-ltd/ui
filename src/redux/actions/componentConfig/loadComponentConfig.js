import _ from 'lodash';
import { LOAD_CONFIG } from '../../actionTypes/componentConfig';
import { initialComponentConfigStates } from '../../reducers/componentConfig/initialState';

const loadComponentConfig = (experimentId, plotUuid, type) => (dispatch) => {
  dispatch({
    type: LOAD_CONFIG,
    payload: {
      experimentId,
      plotUuid,
      type,
      config: _.cloneDeep(initialComponentConfigStates[type]),
    },
  });
};

export default loadComponentConfig;
