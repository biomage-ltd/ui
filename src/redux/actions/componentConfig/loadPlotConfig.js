import _ from 'lodash';
import { LOAD_CONFIG } from '../../actionTypes/componentConfig';
import getApiEndpoint from '../../../utils/apiEndpoint';
import pushNotificationMessage from '../notifications';
import messages from '../../../components/notification/messages';
import { initialPlotConfigStates } from '../../reducers/componentConfig/initialState';

const loadPlotConfig = (experimentId, plotUuid, plotType) => async (dispatch) => {
  try {
    const response = await fetch(
      `${getApiEndpoint()}/v1/experiments/${experimentId}/plots-tables/${plotUuid}`,
    );
    if (response.ok) {
      const data = await response.json();
      dispatch({
        type: LOAD_CONFIG,
        payload: data,
      });
    } else if (response.status === 404) {
      dispatch({
        type: LOAD_CONFIG,
        payload: {
          experimentId,
          plotUuid,
          type: plotType,
          config: _.cloneDeep(initialPlotConfigStates[plotType]),
        },
      });
    } else {
      throw new Error('Server sent back different error or json conversion failed.');
    }
  } catch (e) {
    dispatch(pushNotificationMessage('error', messages.saveCellSets, 10));
  }
};

export default loadPlotConfig;
