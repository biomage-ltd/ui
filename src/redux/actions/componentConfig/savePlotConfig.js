import { SAVE_CONFIG } from '../../actionTypes/componentConfig';
import getApiEndpoint from '../../../utils/apiEndpoint';

const savePlotConfig = (experimentId, plotUuid) => async (dispatch, getState) => {
  // Do not save the 'outstandingChanges' state to the database.
  const { outstandingChanges, ...content } = getState().plots[plotUuid];

  const response = await fetch(
    `${getApiEndpoint()}/v1/experiments/${experimentId}/plots-tables/${plotUuid}`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(content),
    },
  );
  const { lastUpdated } = await response.json();
  dispatch({
    type: SAVE_CONFIG,
    payload:
      { plotUuid, lastUpdated },
  });
};

export default savePlotConfig;
