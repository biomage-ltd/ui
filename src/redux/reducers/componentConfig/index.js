import { UPDATE_CONFIG, SAVE_CONFIG, LOAD_CONFIG } from '../../actionTypes/componentConfig';
import initialState from './initialState';
import loadConfig from './loadConfig';
import updateConfig from './updateConfig';
import saveConfig from './saveConfig';

const plotsReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOAD_CONFIG:
      return loadConfig(state, action);
    case UPDATE_CONFIG:
      return updateConfig(state, action);
    case SAVE_CONFIG:
      return saveConfig(state, action);
    default:
      return state;
  }
};

export default plotsReducer;
