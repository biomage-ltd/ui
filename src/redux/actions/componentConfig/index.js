import loadPlotConfig from './loadPlotConfig';
import loadComponentConfig from './loadComponentConfig';
import updatePlotConfig from './updatePlotConfig';
import savePlotConfig from './savePlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
};
