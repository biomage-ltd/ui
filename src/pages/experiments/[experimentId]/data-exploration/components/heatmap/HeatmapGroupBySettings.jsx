import React, { useEffect, useState, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import _ from 'lodash';
import { updatePlotConfig } from '../../../../../../redux/actions/componentConfig';
import {
  PlusOutlined,
  MinusOutlined,
  UpOutlined,
  DownOutlined,
} from '@ant-design/icons';
import 'antd/dist/antd.css';
import {
  Button, Space, Menu, Dropdown
} from 'antd';

const HeatmapGroupBySettings = () => {
  const dispatch = useDispatch();
  const groupedTracksKeys = useSelector((state) => state.componentConfig.interactiveHeatmap.config.groupedTracks);
  const cellSets = useSelector((state) => state.cellSets);

  const getCellSets = (category) => {
    if (!cellSets || cellSets.loading) {
      return [];
    }

    return cellSets.hierarchy.map(
      ({ key }) => (
        { key, name: cellSets.properties[key].name, type: cellSets.properties[key].type }
      ),
    ).filter(
      ({ type }) => category.includes(type),
    );
  };

  const getCellSetsOrder = () => {
    const allCellSetsGroupBys = getCellSets(['cellSets', 'metadataCategorical']);

    let groupedCellSets = [];

    // from the enabled cell sets keys we get, find their corresponding information
    groupedTracksKeys
      .forEach((trackKey) => {
        const groupBy = allCellSetsGroupBys.find((cellSetGroupBy) => cellSetGroupBy.key === trackKey);
        groupedCellSets.push(groupBy);
      });

    // About the filtering: If we have failed to find some of the groupbys information, 
    // then ignore those (this is useful for groupbys that sometimes dont show up, like 'samples')
    return groupedCellSets.filter((groupedCellSet) => groupedCellSet !== undefined);
  };

  const isInitialRenderRef = useRef(true);
  const [cellSetsOrder, setCellSetsOrder] = useState(getCellSetsOrder());

  useEffect(() => {
    if (isInitialRenderRef.current) {
      isInitialRenderRef.current = false;

      return;
    }

    if (cellSetsOrder.length === 0) {
      return;
    }

    dispatch(
      updatePlotConfig('interactiveHeatmap', {
        groupedTracks: cellSetsOrder.map((cellSet) => cellSet.key)
      })
    );
  }, [cellSetsOrder]);

  const moveUp = (source, id) => {
    const index = source.findIndex((e) => e.key === id);

    const arr = [...source];

    if (index <= 0) {
      return arr;
    }

    const el = arr[index];
    arr[index] = arr[index - 1];
    arr[index - 1] = el;

    return arr;
  };

  const moveDown = (source, id) => {
    const index = source.findIndex((e) => e.key === id);

    const arr = [...source];

    if (index === -1 || index >= source.length - 1) {
      return arr;
    }

    const el = arr[index];
    arr[index] = arr[index + 1];
    arr[index + 1] = el;

    return arr;
  };

  const indexOfCellSet = (cellSet) => {
    return cellSetsOrder.findIndex((elem) => (elem.key === cellSet.key));
  }

  const menu = (
    <Menu>
      {
        getCellSets(['cellSets', 'metadataCategorical'])
          .map((cellSet) => {
            let positionInCellSetOrder = indexOfCellSet(cellSet);

            return (
              <Menu.Item size='small'>
                <div style={{ margin: '0px', width: '100%' }} onClick={(e) => { e.stopPropagation(); }}>
                  <Button
                    shape="square"
                    size='small'
                    style={{ marginRight: '5px' }}
                    icon={positionInCellSetOrder > -1 ? <MinusOutlined /> : <PlusOutlined />}
                    onClick={() => {
                      let newCellSetsOrder = [...cellSetsOrder];
                      if (positionInCellSetOrder > -1) {
                        // If the cell is included in the cellSet, we have to remove it
                        newCellSetsOrder.splice(positionInCellSetOrder, 1);
                      } else {
                        // If the cell is not included in the cellSet, we have to add it
                        newCellSetsOrder.push(cellSet);
                      }

                      setCellSetsOrder(newCellSetsOrder);
                    }}
                  >
                  </Button>
                  {cellSet.name}
                </div>
              </Menu.Item>
            );
          })
      }
    </Menu>
  );

  return (
    <div style={{ padding: '5px' }}>
      <Space direction='vertical'>
        <Dropdown overlay={menu}>
          <div style={{ padding: '7px', border: '1px solid rgb(238,238,238)' }}>
            Select the parameters to group by
            <DownOutlined style={{ marginLeft: '5px' }} />
          </div>
        </Dropdown>

        {
          cellSetsOrder.map((cellSet, i) => (
            <div>
              <Button
                size='small'
                icon={<UpOutlined />}
                shape='circle'
                disabled={i === 0}
                onClick={() => {
                  setCellSetsOrder(moveUp(cellSetsOrder, cellSet.key));
                }}
              />

              <Button
                size='small'
                shape='circle'
                disabled={i === cellSetsOrder.length - 1}
                icon={<DownOutlined />}
                onClick={() => {
                  setCellSetsOrder(moveDown(cellSetsOrder, cellSet.key));
                }}
                style={{ marginRight: '5px' }}
              />

              {cellSet.name}
            </div>
          ))
        }
      </Space>
    </div >
  );

};

HeatmapGroupBySettings.defaultProps = {
};

HeatmapGroupBySettings.propTypes = {
};

export default HeatmapGroupBySettings;