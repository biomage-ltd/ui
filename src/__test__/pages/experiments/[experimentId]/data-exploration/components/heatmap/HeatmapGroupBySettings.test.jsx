import React from 'react';
import { shallow, configure, mount, key } from 'enzyme';
import { Provider } from 'react-redux';
import waitForActions from 'redux-mock-store-await-actions';
import Adapter from 'enzyme-adapter-react-16';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Button, Menu, Dropdown, Space } from 'antd';
import HeatmapGroupBySettings from '../../../../../../../pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapGroupBySettings';
import { UPDATE_CONFIG } from '../../../../../../../redux/actionTypes/componentConfig';

jest.mock('localforage');
jest.mock('../../../../../../../pages/experiments/[experimentId]/data-exploration/components/heatmap/VegaHeatmap');

const mockStore = configureMockStore([thunk]);
configure({ adapter: new Adapter() });

let component;

const initialState = {
  cellSets: {
    hierarchy: [
      {
        key: 'louvain',
        children: [
          {
            key: 'louvain-0',
          },
          {
            key: 'louvain-1',
          },
        ],
      },
      {
        key: 'sample',
        children: [
          {
            key: 'control',
          },
          {
            key: 'treated',
          },
        ],
      },
    ],
    properties: {
      louvain: {
        type: 'cellSets',
        name: 'louvain clusters',
      },
      'louvain-0': {
        name: 'louvain 0',
        cellIds: [5, 6, 7],
      },
      'louvain-1': {
        name: 'louvain 1',
        cellIds: [1, 2, 3],
      },
      sample: {
        type: 'metadataCategorical',
        name: 'Sample',
      },
      control: {
        name: 'control',
        cellIds: [5, 6, 7],
      },
      treated: {
        name: 'treated',
        cellIds: [1, 2, 3],
      },
    },
  },
  cellInfo: {},
  componentConfig: {
    interactiveHeatmap: {
      config: {
        groupedTracks: ['sample'],
        selectedTracks: ['louvain'],
      },
    },
  },
};

describe('HeatmapGroupBySettings', () => {
  afterEach(() => {
    component.unmount();
  });

  it('renders correctly', () => {
    const store = mockStore({
      ...initialState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapGroupBySettings experimentId='123' />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapGroupBySettings').length).toEqual(1);

    // with two items
    const groupByItems = component.find('Item');
    expect(groupByItems.length).toEqual(2);

    // the second item should be samples (the first one is the dropdown)
    expect(groupByItems.at(1).childAt(0).childAt(0).text()).toEqual('Sample');

    // there's a dropdown
    const dropdown = component.find(Dropdown);
    const submenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const subMenuItems = submenu.find('MenuItem');

    // With two items.
    expect(subMenuItems.length).toEqual(2);

    // each of which should have the right names.
    expect(subMenuItems.at(0).childAt(0).text()).toEqual('louvain clusters');
    expect(subMenuItems.at(1).childAt(0).text()).toEqual('Sample');
  });

  test('interacting with the groupby add/remove options will trigger the appropriate actions', async () => {
    const store = mockStore({
      ...initialState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapGroupBySettings experimentId='123' width={200} height={200} />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapGroupBySettings').length).toEqual(1);

    const dropdown = component.find(Dropdown);
    const submenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const subMenuItems = submenu.find('MenuItem');

    // When the other group by is clicked...
    const buttons = subMenuItems.find(Button);
    buttons.at(0).simulate('click');
    component.update();

    await waitForActions(store, [UPDATE_CONFIG]);

    // The store should update.
    expect(store.getActions().length).toEqual(1);
    const firstAction = store.getActions()[0];
    expect(firstAction.type).toBe(UPDATE_CONFIG);
    expect(firstAction).toMatchSnapshot();
    store.clearActions();
    expect(store.getActions().length).toEqual(0);

    // there should be 3 group by items
    const groupByItems = component.find('Item');
    expect(groupByItems.length).toEqual(3);

    // the second item should be samples (the first one is the dropdown)
    expect(groupByItems.at(1).childAt(0).childAt(0).text()).toEqual('Sample');
    // the third item should be louvain (the first one is the dropdown)
    expect(groupByItems.at(2).childAt(0).childAt(0).text()).toEqual('louvain clusters');

    // when the groupby is clicked again
    buttons.at(0).simulate('click');
    component.update();
    
    await waitForActions(store, [UPDATE_CONFIG]);

    // The store should update.
    expect(store.getActions().length).toEqual(1);
    const secondAction = store.getActions()[0];

    expect(secondAction.type).toBe(UPDATE_CONFIG);
    expect(secondAction).toMatchSnapshot();

    // there should be 2 group by items
    const updatedGroupByItems = component.find('Item');
    expect(groupByItems.length).toEqual(3);

    // and the second item should be samples (the first one is the dropdown)
    expect(groupByItems.at(1).childAt(0).childAt(0).text()).toEqual('Sample');
  });

  test('interacting with the groupby reorder options will trigger the appropriate actions', async () => {
    const store = mockStore({
      ...initialState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapGroupBySettings experimentId='123' width={200} height={200} />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapGroupBySettings').length).toEqual(1);

    const dropdown = component.find(Dropdown);
    const submenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const subMenuItems = submenu.find('MenuItem');

    // Add a louvain group by
    const addButtons = subMenuItems.find(Button);
    addButtons.at(0).simulate('click');
    component.update();

    // there should be 3 group by items
    const groupByItems = component.find('Item');
    expect(groupByItems.length).toEqual(3);

    // clear store from this action (not being tested here)
    await waitForActions(store, [UPDATE_CONFIG]);
    store.clearActions();

    // the second item should be samples (the first one is the dropdown)
    expect(groupByItems.at(1).childAt(0).childAt(0).text()).toEqual('Sample');
    // the third item should be louvain (the first one is the dropdown)
    expect(groupByItems.at(2).childAt(0).childAt(0).text()).toEqual('louvain clusters');

    // after a click for sample to go down
    const reorderButtonsSample = groupByItems.at(1).find('Button');
    console.log('reorderButtonsSample');
    console.log(reorderButtonsSample.debug());
    reorderButtonsSample.at(1).simulate('click');

    // the store should update.
    await waitForActions(store, [UPDATE_CONFIG]);
    expect(store.getActions().length).toEqual(1);
    const secondAction = store.getActions()[0];
    expect(secondAction).toMatchSnapshot();
  });
});
