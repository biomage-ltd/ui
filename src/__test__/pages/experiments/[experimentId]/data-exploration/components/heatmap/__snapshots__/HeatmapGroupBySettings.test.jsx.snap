// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapGroupBySettings clicking on the other button will trigger the appropriate action 1`] = `
Object {
  "payload": Object {
    "configChange": Object {
      "groupedTrack": "louvain",
    },
    "plotUuid": "interactiveHeatmap",
  },
  "type": "componentConfig/update",
}
`;
